File: go-service/internal/interface/http/error_handler.go
package http

import (
	"errors"
	"net/http"

	"github.com/gofiber/fiber/v2"

	"github.com/example/interview-stack/go-service/internal/application/dto"
	"github.com/example/interview-stack/go-service/internal/domain"
)

func handleError(ctx *fiber.Ctx, err error) error {
	var fiberErr *fiber.Error
	if errors.As(err, &fiberErr) {
		return ctx.Status(fiberErr.Code).JSON(dto.ErrorResponse{Error: fiberErr.Message})
	}
	var domainErr *domain.Error
	if errors.As(err, &domainErr) {
		switch domainErr.Code {
		case domain.ErrorCodeValidation:
			return ctx.Status(http.StatusBadRequest).JSON(dto.ErrorResponse{Error: domainErr.Message})
		case domain.ErrorCodeNotFound:
			return ctx.Status(http.StatusNotFound).JSON(dto.ErrorResponse{Error: domainErr.Message})
		default:
			return ctx.Status(http.StatusInternalServerError).JSON(dto.ErrorResponse{Error: domainErr.Message})
		}
	}
	return ctx.Status(http.StatusInternalServerError).JSON(dto.ErrorResponse{Error: "internal server error"})
}

File: node-service/internal/interface/http/errorHandler.js
const DomainError = require('../../domain/errors');

function mapError(err, res) {
  if (err instanceof DomainError) {
    if (err.code === 'validation') {
      return res.status(400).json({ error: err.message });
    }
    if (err.code === 'not_found') {
      return res.status(404).json({ error: err.message });
    }
    return res.status(500).json({ error: err.message });
  }

  if (err && err.statusCode) {
    return res.status(err.statusCode).json({ error: err.message });
  }

  return res.status(500).json({ error: 'internal server error' });
}

module.exports = mapError;
